Incremental CAgg refreshes for tiered data (#42)
timescale/timescaledb-extras: utils/osm_incremental_cagg_refresh/README.md

```

## Advanced Usecase

1. You can use the schedule_osm_cagg_refresh procedure to generate refresh ranges for multiple continuous aggregates.

```sql
CALL _timescaledb_additional.schedule_osm_cagg_refresh(); -- All CAggs with tiered data
```

```sql
CALL _timescaledb_additional.schedule_osm_cagg_refresh(schema_mask=>'hf_schema'); -- All CAggs with tiered data under hf_schema 
```

```sql
CALL _timescaledb_additional.schedule_osm_cagg_refresh( name_mask=>'cagg_data%'); -- All CAggs with name beginning with cagg_data, that appears in any schema  
```

```sql
CALL _timescaledb_additional.schedule_osm_cagg_refresh( schema_mask=>'hf_schema%', name_mask=>'cagg_data%'); -- All CAggs with name beginning with cagg_data under schemas that begin with the name hf_schema.  
```

2. Multiple continuous aggregates can be refreshed in parallel. You can create multiple jobs like so:

```sql
SELECT
    add_job(
        '_timescaledb_additional.task_refresh_continuous_aggregate_incremental_runner',
        schedule_interval => '5 seconds',
        config => '{"enable_tiered_reads": true, "max_runtime": "5 minutes"}'
    )
FROM
    generate_series(1, 4);
```

> Every job is run by a background worker. Please keep you max background worker settings and database load in mind while adding jobs so that you do not exhaust DB resources.

Author: ```suggestion
Every job is executed by a background worker. Please keep you max background worker settings and database load in mind while adding jobs so that you do not exhaust DB resources.
```

# Incremental Continuous Aggregate Refresh on Tiered Data

## Introduction

> This directory has a set of SQL procedures to create/refresh a continuous aggregates from tiered data on S3.

Reviewer: ```suggestion
This directory has a set of SQL procedures to create/refresh a continuous aggregates from tiered data on S3, only available on Timescale Cloud.
```

## Producer
The procedure `_timescaledb_additional.schedule_osm_cagg_refresh` finds the time range of the tiered data for the hypertable (corresponding to the Continuous Aggregate) and splits the range into a set of time intervals. These intervals are added to the `_timescaledb_additional.incremental_continuous_aggregate_refreshes` table.

## Consumer
The procedure `_timescaledb_additional.task_refresh_continuous_aggregate_incremental_runner` picks the time intervals from the `_timescaledb_additional.incremental_continuous_aggregate_refreshes` table and calls `refresh_continuous_aggregate` procedure on these intervals until the list is exhausted.

## How do I run this?

### 1. Produce ranges to be refreshed:
```sql
CALL _timescaledb_additional.schedule_osm_cagg_refresh(); -- All CAggs with tiered data
```
**NOTE**: You can specify an Continuous Aggregate to schedule the range to be refreshed passing the name as the first parameter. I.e.:
```sql
CALL _timescaledb_additional.schedule_osm_cagg_refresh('<name of continuous aggregate>');
```

> ### 2. Add a job to consume (aka refresh) the produced ranges:

Reviewer: ```suggestion
### 2. Add a consumer job to refresh the produced ranges:
```

```sql
SELECT
    add_job(
        '_timescaledb_additional.task_refresh_continuous_aggregate_incremental_runner',
        schedule_interval => '5 seconds',
        config => '{"enable_tiered_reads": true, "max_runtime": "5 minutes"}'
    );
```

> > **IMPORTANT NOTES**:
> > 1. The `enable_tiered_reads = true` config is necessary if your default DB settings for `timescaleb_osm.enable_tiered_reads` GUC is `false`.
> > 2. A Continuous Aggregate will be refreshed by exactly 1 (one) job using this framework even if you choose to add multiple jobs to run the `task_refresh_continuous_aggregate_incremental_runner` procedure.

Reviewer: ```suggestion
```

    );
```

This call adds a job that runs in the background and refreshes the continuous aggregate.

**NOTE**
The `enable_tiered_reads = true` config is necessary if the default DB settings for `timescaleb_osm.enable_tiered_reads` GUC is `false`.

### 3. Monitoring

The `_timescaledb_additional.osm_incremental_refresh_status` view can be used to monitor progress of the refresh execution:
```sql
tsdb=> SELECT * FROM _timescaledb_additional.osm_incremental_refresh_status;
       continuous_aggregate       | not started | started | finished 
----------------------------------+-------------+---------+----------
 hf_dss.g_data_2a_43_1minute      |         256 |       0 |        3
 hf_dss.g_data_2a_43_1hour        |         255 |       1 |        3
 hf_dss.g_data_fenceline_1minute  |         267 |       0 |        3
 hf_dss.g_data_fenceline_1hour    |         267 |       0 |        3
 hf_dss.g_data_16_41_1minute      |         258 |       1 |       27
 hf_dss.g_data_16_41_1hour        |         258 |       0 |       28
 hf_dss."g_data_2b-21-35_1minute" |         260 |       1 |        3
 hf_dss."g_data_2b-21-35_1hour"   |         260 |       1 |        3

> (8 rows)

Author: Why did u removed the entries and keep only one?

```

`finished` shows the number of refresh time intervals that have been completed, `not started` shows how many are remaining.

The `_timescaledb_additional.job_cagg_refresh_status` view can be used to check the status of the job executions (consumers):
```sql
tsdb=> SELECT * FROM _timescaledb_additional.job_cagg_refresh_status;
    clock_timestamp     |  pid   |  wait_event  |                        application_name                        | xact_age | backend_age 
------------------------+--------+--------------+----------------------------------------------------------------+----------+-------------
 2025-01-04 14:31:03+00 | 153087 |              | hf_dss.g_data_fenceline_1hour refresh 2021-01-19 2021-02-18    | 00:00:10 | 00:00:44
 2025-01-04 14:31:03+00 | 153088 | DataFileRead | hf_dss.g_data_fenceline_1minute refresh 2021-01-19 2021-02-18  | 00:00:20 | 00:00:44
 2025-01-04 14:31:03+00 | 153089 |              | hf_dss.g_data_16_41_1minute refresh 2021-02-11 2021-03-13      | 00:00:10 | 00:00:44
 2025-01-04 14:31:03+00 | 153090 |              | hf_dss."g_data_2b-21-35_1minute" refresh 2021-01-18 2021-02-17 | 00:00:24 | 00:00:44

> (4 rows)

Author: Why did u removed the entries and keep only one?
